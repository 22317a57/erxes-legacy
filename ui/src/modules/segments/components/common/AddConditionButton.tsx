import Button from 'modules/common/components/Button';
import Icon from 'modules/common/components/Icon';
import { __ } from 'modules/common/utils';
import React from 'react';
import OverlayTrigger from 'react-bootstrap/OverlayTrigger';
import Popover from 'react-bootstrap/Popover';
import { ISegmentCondition, ISegmentField } from '../../types';
import { FieldType, PopoverList } from '../styles';

type Props = {
  fields: ISegmentField[];
  events: string[];
  addCondition: (condition: ISegmentCondition) => void;
};

class AddConditionButton extends React.Component<Props> {
  private overlayTrigger;

  addCondition = (type: string) => {
    this.props.addCondition({
      key: Math.random().toString(),
      propertyName: '',
      propertyValue: '',
      propertyOperator: '',
      type,
    });

    this.overlayTrigger.hide();
  };

  renderPopover() {
    const addPropertyCondition = () => {
      return this.addCondition('property');
    }

    const addEventCondition = () => {
      return this.addCondition('event');
    }

    return (
      <Popover id="condition-popover">
        <Popover.Title as="h3">{__('Select a field')}</Popover.Title>
        <Popover.Content>
          <PopoverList onClick={addPropertyCondition}>
            <FieldType>
              {__("Fields")}
              <Icon icon="information" />
            </FieldType>
          </PopoverList>

          <PopoverList onClick={addEventCondition}>
            <FieldType>
              {__("Events")}
              <Icon icon="settings-3" />
            </FieldType>
          </PopoverList>
        </Popover.Content>
      </Popover>
    );
  }

  render() {
    return (
      <OverlayTrigger
        ref={overlayTrigger => {
          this.overlayTrigger = overlayTrigger;
        }}
        trigger="click"
        placement="bottom"
        overlay={this.renderPopover()}
        container={this}
        rootClose={true}
      >
        <Button btnStyle="success" icon="add">
          Add a condition
        </Button>
      </OverlayTrigger>
    );
  }
}

export default AddConditionButton;
