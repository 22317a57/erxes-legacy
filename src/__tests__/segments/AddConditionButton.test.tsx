import { mount, shallow } from 'enzyme';
import * as React from 'react';
import * as renderer from 'react-test-renderer';

import AddConditionButton from '../../modules/segments/components/AddConditionButton';

describe('AddConditionButton component', () => {
  const defaultProps = {
    fields: [],
    addCondition: (condition: { _id: string }) => null
  };

  test('renders AddConditionButton successfully', () => {
    shallow(<AddConditionButton {...defaultProps} />);
  });

  test('renders successfully with default value', () => {
    const wrapper = mount(<AddConditionButton {...defaultProps} />);
    const props = wrapper.props();

    expect(props).toMatchObject(defaultProps);
  });

  test('snapshot matches', () => {
    const rendered = renderer
      .create(<AddConditionButton {...defaultProps} />)
      .toJSON();

    expect(rendered).toMatchSnapshot();
  });
});
