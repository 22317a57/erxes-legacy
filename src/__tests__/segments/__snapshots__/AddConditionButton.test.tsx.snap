// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddConditionButton component snapshot matches 1`] = `
<button
  className="sc-ifAKCX cQdFDO"
  icon="add"
  onClick={[Function]}
  type="button"
>
  <i
    className="icon-add sc-bwzfXH hpxbGw"
    icon="add"
  />
  <span>
    Add a condition
  </span>
</button>
`;
